domain/consensus/utils/consensushashing/calculate_signature_hash.go

```diff
@@ -0,0 +1,196 @@
+
+	if prevScriptPublicKey.Version > constants.MaxScriptPublicKeyVersion {
+		return nil, errors.Errorf("Script version is unkown.")
+	}
+
+	return calculateSignatureHash(tx, inputIndex, txIn, prevScriptPublicKey, hashType, reusedValues)
+}
+
+func calculateSignatureHash(tx *externalapi.DomainTransaction, inputIndex int, txIn *externalapi.DomainTransactionInput,
+	prevScriptPublicKey *externalapi.ScriptPublicKey, hashType SigHashType, reusedValues *SighashReusedValues) (
+	*externalapi.DomainHash, error) {
+
+	hashWriter := hashes.NewTransactionSigningHashWriter()
+	infallibleWriteElement(hashWriter, tx.Version)
+
+	previousOutputsHash := getPreviousOutputsHash(tx, hashType, reusedValues)
+	infallibleWriteElement(hashWriter, previousOutputsHash)
+
+	sequencesHash := getSequencesHash(tx, hashType, reusedValues)
+	infallibleWriteElement(hashWriter, sequencesHash)
+
+	hashOutpoint(hashWriter, txIn)
+
+	infallibleWriteElement(hashWriter, prevScriptPublicKey.Version)
+	infallibleWriteElement(hashWriter, prevScriptPublicKey.Script)
+
+	infallibleWriteElement(hashWriter, txIn.UTXOEntry.Amount())
+
+	infallibleWriteElement(hashWriter, txIn.Sequence)
+
+	outputsHash := getOutputsHash(tx, inputIndex, hashType, reusedValues)
+	infallibleWriteElement(hashWriter, outputsHash)
+
+	infallibleWriteElement(hashWriter, tx.LockTime)
+
+	infallibleWriteElement(hashWriter, tx.SubnetworkID)
+	infallibleWriteElement(hashWriter, tx.Gas)
+
+	payloadHash := getPayloadHash(tx, reusedValues)
+	infallibleWriteElement(hashWriter, payloadHash)
+
+	infallibleWriteElement(hashWriter, uint32(hashType))
+
+	return hashWriter.Finalize(), nil
+}
+
+func getPreviousOutputsHash(tx *externalapi.DomainTransaction, hashType SigHashType, reusedValues *SighashReusedValues) *externalapi.DomainHash {
+	if hashType.isSigHashAnyOneCanPay() {
+		return externalapi.NewZeroHash()
+	}
+
+	if reusedValues.previousOutputsHash == nil {
+		hashWriter := hashes.NewTransactionSigningHashWriter()
+		for _, txIn := range tx.Inputs {
+			hashOutpoint(hashWriter, txIn)
+		}
+		reusedValues.previousOutputsHash = hashWriter.Finalize()
+	}
+
+	return reusedValues.previousOutputsHash
+}
+
+func getSequencesHash(tx *externalapi.DomainTransaction, hashType SigHashType, reusedValues *SighashReusedValues) *externalapi.DomainHash {
+	if hashType.isSigHashSingle() || hashType.isSigHashAnyOneCanPay() || hashType.isSigHashNone() {
+		return externalapi.NewZeroHash()
+	}
+
+	if reusedValues.sequencesHash == nil {
+		hashWriter := hashes.NewTransactionSigningHashWriter()
+		for _, txIn := range tx.Inputs {
+			infallibleWriteElement(hashWriter, txIn.Sequence)
+		}
+		reusedValues.sequencesHash = hashWriter.Finalize()
+	}
+
+	return reusedValues.sequencesHash
+}
+
+func getOutputsHash(tx *externalapi.DomainTransaction, inputIndex int, hashType SigHashType, reusedValues *SighashReusedValues) *externalapi.DomainHash {
+	// SigHashNone: return zero-hash
+	if hashType.isSigHashNone() {
+		return externalapi.NewZeroHash()
+	}
```

Comment:
I think there's a meaningful difference between this and the bitcoin approach.
the difference is how do we handle unknown sighashes, in bitcoin unknown hash types will be all zeros here, and in this code `SigHashAll` will be the default.
I'm not sure which is better and what kinds of validation happen to the sighash before this function (maybe only valid known sighashes can enter this function)

Reply:
Same for `getSequencesHash`

Reply:
I've added a validation in the beggining of CalculateSignatureHash that makes sure that the SigHashType passed is a standard one, so we are not going to have a default (it will error out if you pass 0)